File: src/Stack/Stack.c
/*
 * @Author: xixi_
 * @Date: 2024-10-20 18:40:49
 * @LastEditors: xixi_
 * @LastEditTime: 2024-11-11 10:52:59
 * @FilePath: /FHMF/src/Modules/xixi/src/Stack/Stack.c
 * Copyright (c) 2023-2024 by xixi_ , All Rights Reserved.
 */

#include "Stack.h"

void XIXI_StackInit(ThisStack *ThisStack)
{
    ThisStack->ThisTop = -1; /* 栈顶指针置为-1 */
}

int XIXI_StackIsEmpty(ThisStack *ThisStack)
{
    return ThisStack->ThisTop == -1; /* 判断栈顶是否等于-1即可 */
}

int XIXI_StackPush(ThisStack *ThisStack, const char *ThisVal)
{
    if (ThisStack->ThisTop == STACK_MAX_SIZE - 1)
    {
        return 0; /* 栈满,推入失败 */
    }
    if (!ThisVal)
    {
        return 0; /* 空字符,推入失败 */
    }
    char *ThisTmpVal = strdup(ThisVal);
    if (ThisTmpVal == NULL)
    {
        return 0; /* 内存分配失败,推入失败 */
    }
    ThisStack->ThisCharArr[++(ThisStack->ThisTop)] = ThisTmpVal;

    return 1;
}

int XIXI_StackAppendVal(ThisStack *ThisStack, const char *ThisVal)
{
    if (XIXI_StackIsEmpty(ThisStack))
    {
        XIXI_StackPush(ThisStack, ThisVal); /* 如果栈为空，直接推入 */
        return 1;
    }
    char *TopStr = XIXI_StackPeek(ThisStack); /* 获取栈顶当前字符串并扩展内存 */
    if (!ThisVal || !TopStr)
    {
        return 0;
    }
    size_t newLength = strlen(TopStr) + strlen(ThisVal) + 1; /* 新字符串的长度 */
    char *NewChar = (char *)realloc(TopStr, newLength);      /* 重分配大小 */
    if (NewChar == NULL)
    {
        return 0; /* 内存分配失败 */
    }
    strcat(NewChar, ThisVal);                             /* 连接字符串 */
    ThisStack->ThisCharArr[ThisStack->ThisTop] = NewChar; /* 更新栈顶指针 */

    return 1;
}

char *XIXI_StackPop(ThisStack *ThisStack)
{
    if (XIXI_StackIsEmpty(ThisStack))
    {
        return NULL; /* 栈空,不能出栈 */
    }
    return ThisStack->ThisCharArr[(ThisStack->ThisTop)--]; /* 返回栈顶元素并更新栈顶指针 */
}

char *XIXI_StackPeek(ThisStack *ThisStack)
{
    if (XIXI_StackIsEmpty(ThisStack))
    {
        return NULL; /* 栈空,无法查看 */
    }
    return ThisStack->ThisCharArr[ThisStack->ThisTop]; /* 返回栈顶元素,不更新栈顶指针 */
}

void XIXI_StackPrint(ThisStack *ThisStack)
{
    if (XIXI_StackIsEmpty(ThisStack))
    {
        printf("Error: Stack is empty.\n");
        return;
    }
    for (int i = 0; i <= ThisStack->ThisTop; i++)
    {
        printf("%s\n", ThisStack->ThisCharArr[i]);
    }
}

void XIXI_StackFree(ThisStack *ThisStack)
{
    for (int i = 0; i <= ThisStack->ThisTop; i++)
    {
        free(ThisStack->ThisCharArr[i]);  /* 释放每一个串的内存 */
        ThisStack->ThisCharArr[i] = NULL; /* 避免悬挂指针 */
    }
}

File: src/Stack/Stack.lua
--[[
Author: xixi_
Date: 2024-10-27 17:48:03
LastEditors: xixi_
LastEditTime: 2024-11-11 11:58:33
FilePath: /FHMF/src/Modules/xixi/src/Stack/Stack.lua
Copyright (c) 2020-2024 by xixi_ , All Rights Reserved.
--]]
--------------------------------------------------------------------------------------------------------------------------
Stack = {}; --[[ 栈类 ]]
Stack.__index = Stack; --[[ 元表 ]]
--------------------------------------------------------------------------------------------------------------------------
--[[ 栈构造函数 ]]
function Stack:new()
    local instance = {
        ThisData = {},
        ThisTop = 0
    };
    setmetatable(instance, Stack);
    return instance;
end
--------------------------------------------------------------------------------------------------------------------------
--[[ 判断栈是否空 ]]
function Stack:IsEmpty()
    return self.ThisTop == 0;
end
--------------------------------------------------------------------------------------------------------------------------
--[[ 将`value`推进栈 ]]
function Stack:Push(value)
    self.ThisTop = self.ThisTop + 1;
    self.ThisData[self.ThisTop] = value;
end
--------------------------------------------------------------------------------------------------------------------------
--[[ 向栈顶追加元素 ]]
function Stack:AppendVal(value)
    if (self:IsEmpty()) then
        self:Push(value); --[[ 栈空,直接推入 ]]
        return;
    end
    self.ThisData[self.ThisTop] = self.ThisData[self.ThisTop] .. value; --[[ 栈非空,在末尾追加 ]]
end
--------------------------------------------------------------------------------------------------------------------------
--[[ 弹出栈顶元素 ]]
function Stack:Pop()
    if (self:IsEmpty()) then
        return nil; --[[ 栈为空，返回 nil ]]
    end
    local value = self.ThisData[self.ThisTop];
    self.ThisTop = self.ThisTop - 1;
    return value;
end
--------------------------------------------------------------------------------------------------------------------------
--[[ 返回栈顶元素 ]]
function Stack:Peek()
    if (self:IsEmpty()) then
        return nil; --[[ 栈为空，返回 nil ]]
    end
    return self.ThisData[self.ThisTop];
end
--------------------------------------------------------------------------------------------------------------------------
--[[ 打印栈 ]]
function Stack:Print()
    for i = 1, self.ThisTop, 1 do
        print(self.ThisData[i]);
    end
end
--------------------------------------------------------------------------------------------------------------------------
--[[ 使用示例 ]]
local Stack = Stack:new();
Stack:Push("Fuck YanZiHao");
Stack:Pop();
Stack:AppendVal(" Fuck YanJianKang");
Stack:Print();
--------------------------------------------------------------------------------------------------------------------------
